File: CS261/Asgmt03/Library.cpp
#include "memoryleakdetect.h"		// needs to be first #include in the .cpp file
#include <cstdarg>					// support macros for vararg
#include <iostream>
#include "Library.h"
#include "Book.h"
#include "MusicAlbum.h"
#include "Movie.h"

// general functions

Library::Library() :
	setBooks(),
	setMovies(),
	setAlbums()
	
{
}

void Library::addKeywordsForItem(Item* item, int nKeywords, ...)
{
	// the code in this function demonstrates how to handle a vararg in C++

	va_list		keywords;
	char		*keyword;

	va_start(keywords, nKeywords);
	for (int i = 0; i < nKeywords; i++) {
		keyword = va_arg(keywords, char*);
		// do something with each keyword

		string newKeyword = keyword;
		item->addKeyword(newKeyword);
		}
	va_end(keywords);
	item->sortKeywords();
}

ItemSet* Library::itemsForKeyword(const string keyword) const
{
	ItemSet* tmpSet = new ItemSet();

	// Books
	ItemSet::iterator b_it = setBooks.begin();
	
	for (b_it; b_it != setBooks.end(); ++b_it) {
		vector<string> tmpKeywords;
		tmpKeywords = b_it->getPtr()->getKeywords();

		vector<string>::iterator iter;
		for (iter = tmpKeywords.begin(); iter != tmpKeywords.end(); ++iter) {
			if (*iter == keyword) {
				tmpSet->insert(b_it->getPtr());
			}
		}
	}

	// Movies

	ItemSet::iterator m_it = setMovies.begin();
	for (m_it; m_it != setMovies.end(); ++m_it) {
		vector<string> tmpKeywords;
		tmpKeywords = m_it->getPtr()->getKeywords();

		vector<string>::iterator iter;
		for (iter = tmpKeywords.begin(); iter != tmpKeywords.end(); ++iter) {
			if (*iter == keyword) {
				tmpSet->insert(m_it->getPtr());
			}
		}
	}

	// Music Albums

	ItemSet::iterator ma_it = setAlbums.begin();
	for (ma_it; ma_it != setAlbums.end(); ++ma_it) {
		vector<string> tmpKeywords;
		tmpKeywords = ma_it->getPtr()->getKeywords();

		vector<string>::iterator iter;
		for (iter = tmpKeywords.begin(); iter != tmpKeywords.end(); ++iter) {
			if (*iter == keyword) {
				tmpSet->insert(ma_it->getPtr());
			}
		}
	}
	
	return tmpSet;
}

// book-related functions

Item* Library::addBook(string title, string author, int nPages)
{
	Item* book = new Book(title, author, nPages);
	setBooks.insert(book);
	return book;
}

ItemSet* Library::booksByAuthor(const string author) const
{
	ItemSet* tmpSet = new ItemSet();
	ItemSet::iterator it = setBooks.begin();
	for (it; it != setBooks.end(); ++it) {
		if (it->getPtr()->getArtist() == author) {
			tmpSet->insert(it->getPtr());
		}
	}
	return tmpSet;

}

ItemSet* Library::books() 
{
	ItemSet* tmpSet = new ItemSet();
	ItemSet::iterator it = setBooks.begin();
	for (it; it != setBooks.end(); ++it) {
		tmpSet->insert(it->getPtr());
	}
	
	return tmpSet;
}

// music-related functions

Item* Library::addMusicAlbum(const string title, const string band, const int nSongs)
{
	Item* album = new MusicAlbum(title, band, nSongs);
	setAlbums.insert(album);
	return album;
}

void Library::addBandMembers(Item* const musicAlbum, const int nBandMembers, ...)
{

	va_list		keywords;
	char		*keyword;

	va_start(keywords, nBandMembers);
	for (int i = 0; i < nBandMembers; i++) {
		keyword = va_arg(keywords, char*);
		// do something with each keyword

		string newMember = keyword;
		dynamic_cast<MusicAlbum*>(musicAlbum)->addMember(newMember);
	}
	va_end(keywords);
	// sort members
	//dynamic_cast<MusicAlbum*>(musicAlbum)->sortMembers();
}

ItemSet* Library::musicByBand(const string band) const
{
	ItemSet* tmpSet = new ItemSet();
	ItemSet::iterator it = setAlbums.begin();
	for (it; it != setAlbums.end(); ++it) {
		if (it->getPtr()->getArtist() == band) {
			tmpSet->insert(it->getPtr());
		}
	}
	return tmpSet;

}

ItemSet* Library::musicByMusician(const string musician) const
{
	ItemSet* tmpSet = new ItemSet();
	ItemSet::iterator it = setAlbums.begin();
	for (it; it != setAlbums.end(); ++it) {
		set<string> tmpKeywords;
		tmpKeywords = dynamic_cast<MusicAlbum*>(it->getPtr())->getMembers();
		
		set<string>::iterator iter = tmpKeywords.begin();
		for (iter; iter != tmpKeywords.end(); ++iter) {
			if (*iter == musician) {
				tmpSet->insert(it->getPtr());
			}
		}
	}
	return tmpSet;
}

ItemSet* Library::musicAlbums() 
{
	ItemSet* tmpSet = new ItemSet();
	ItemSet::iterator it = setAlbums.begin();
	for (it; it != setAlbums.end(); ++it) {
		tmpSet->insert(it->getPtr());
	}
	return tmpSet;
}

// movie-related functions

Item* Library::addMovie(const string title, const string director, const int nScenes)
{
	Item* movie = new Movie(title, director, nScenes);
	setMovies.insert(movie);
	return movie;
}

void Library::addCastMembers(Item* const movie, const int nCastMembers, ...)
{
	va_list		keywords;
	char		*keyword;

	va_start(keywords, nCastMembers);
	for (int i = 0; i < nCastMembers; i++) {
		keyword = va_arg(keywords, char*);
		// do something with each keyword

		string newMember = keyword;
		dynamic_cast<Movie*>(movie)->addMember(newMember);
	}
	va_end(keywords);
}

ItemSet* Library::moviesByDirector(const string director) const
{
	ItemSet* tmpSet = new ItemSet();
	ItemSet::iterator it = setMovies.begin();
	for (it; it != setMovies.end(); ++it) {
		if (it->getPtr()->getArtist() == director) {
			tmpSet->insert(it->getPtr());
		}
	}
	return tmpSet;

}

ItemSet* Library::moviesByActor(const string actor) const
{
	ItemSet* tmpSet = new ItemSet();
	ItemSet::iterator it = setMovies.begin();
	for (it; it != setMovies.end(); ++it) {
		set<string> tmpKeywords;
		tmpKeywords = dynamic_cast<Movie*>(it->getPtr())->getCast();

		set<string>::iterator iter;
		for (iter = tmpKeywords.begin(); iter != tmpKeywords.end(); ++iter) {
			if (*iter == actor) {
				tmpSet->insert(it->getPtr());
			}
		}
	}
	return tmpSet;
}

ItemSet* Library::movies() const
{
	ItemSet* tmpSet = new ItemSet();
	ItemSet::iterator it = setMovies.begin();
	for (it; it != setMovies.end(); ++it) {
		tmpSet->insert(it->getPtr());
	}
	return tmpSet;

}

static void deleteMapContents(StringToItemSetMap& s2ism)
{
	// your code here
	// I didn't use a map
}

static void deleteItemSetContents(ItemSet& itemSet)
{
	// your code here
	itemSet.clear();
}

Library::~Library()
{
	// your code here
	setBooks.clear();
	setMovies.clear();
	setAlbums.clear();
}

File: CS261/Asgmt02/library/Library.java
package library;
import java.util.ArrayList;

import java.io.PrintStream;
import java.util.Collection;
import java.util.Collections;
import java.util.List;

public class Library
{
	// general methods
	private ArrayList<Item> items;
	
	//constructor
	public Library(){
		this.items = new ArrayList<>();
	}
	
	// returns all of the items which have the specified keyword
	public ArrayList<Item> itemsForKeyword(String keyword)
	{
		ArrayList<String> stringByKeyword = new ArrayList<>();
		ArrayList<Item> itemByKeyword = new ArrayList<>();
		
		for(Item item: items){
			if(item.getKeywords().contains(keyword)){
				stringByKeyword.add(item.getTitle());
			}
		}
		Collections.sort(stringByKeyword);
		
		for(String thisTitle: stringByKeyword){
			for(Item thisItem: items){
				if((thisItem).getTitle().equals(thisTitle)){
					itemByKeyword.add(thisItem);
				}
				
			}
		}
		return itemByKeyword;
	}
	
	// print an item from this library to the output stream provided
	public void printItem(PrintStream out, Item item)
	{
		if(item instanceof Book){
			out.println((Book) item);
		}
		else if(item instanceof Movie){
			out.println((Movie) item);
		} else {
			out.println((MusicAlbum) item);
		}
	}
	
	// book-related methods
	
	// adds a book to the library
	// NOTE: THIS METHOD ORIGINALLY WAS TYPE ITEM, NOT TYPE VOID. (NOW IVE CHANGED IT BACK)
	public Item addBook(String title, String author, int nPages, String... keywords)
	{
		Item newItem = new Book(title, author, nPages, keywords);
		items.add(newItem);
		return newItem;
	}
	
	// removes a book from the library
	public boolean removeBook(String title)
	{
		for(Item item: items){
			if(item instanceof Book){
				if(((Book) item).getTitle().equals(title)){
					items.remove(item);
					return true;
				}
			}
		}
		return false;
	}
	
	// returns all of the books by the specified author
	public ArrayList<Item> booksByAuthor(String author)
	{
		ArrayList<String> stringTitlesByAuthor = new ArrayList<>();
		ArrayList<Item> itemTitlesByAuthor = new ArrayList<>();
		
		for(Item item: items){
			if(item instanceof Book){
				if(((Book) item).getAuthor().equals(author)){
					stringTitlesByAuthor.add(item.getTitle());
				}
			}
		}
		
		Collections.sort(stringTitlesByAuthor);
		for(String thisTitle: stringTitlesByAuthor){
			for(Item thisItem: items){
				if(thisItem instanceof Book){
					if(((Book) thisItem).getTitle().equals(thisTitle)){
						itemTitlesByAuthor.add(thisItem);
					}
				}
			}
		}
		return itemTitlesByAuthor;
	}
	
	// returns all of the books in the library
	public ArrayList<Item> books()
	{
		ArrayList<String> stringBooks = new ArrayList<>();
		ArrayList<Item> itemBooks = new ArrayList<>();
		
		for(Item item: items){
			if(item instanceof Book){
				stringBooks.add(item.getTitle());
			}
		}
		Collections.sort(stringBooks);
		
		for(String thisTitle: stringBooks){
			for(Item thisItem: items){
				if(thisItem instanceof Book){
					if(((Book) thisItem).getTitle().equals(thisTitle)){
						itemBooks.add(thisItem);
					}
				}
			}
		}
		return itemBooks;
	}
	
	// music-related methods
	
	// adds a music album to the library
	public Item addMusicAlbum(String title, String band, int nSongs, String... keywords)
	{
		Item newItem = new MusicAlbum(title, band, nSongs, keywords);
		items.add(newItem);
		return newItem;
	}

	// adds the specified band members to a music album
	public void addBandMembers(Item album, String... members)
	{
		((MusicAlbum) album).setMembers(members);
	}
	
	// removes a music album from the library
	public boolean removeMusicAlbum(String title)
	{
		for(Item item: items){
			if(item instanceof MusicAlbum){
				if(item.getTitle().equals(title)){
					items.remove(item);
					return true;
				}
			}
		}
		return false;
	}

	// returns all of the music albums by the specified band
	public ArrayList<Item> musicByBand(String band)
	{
		ArrayList<String> stringAllAlbumsByBand = new ArrayList<>();
		ArrayList<Item> itemAllAlbumsByBand = new ArrayList<>();
		
		for(Item item: items){
			if(item instanceof MusicAlbum){
				if(((MusicAlbum) item).getBand().equals(band)){
					stringAllAlbumsByBand.add(item.getTitle());
				}
			}
		}
		
		Collections.sort(stringAllAlbumsByBand);
		for(String thisTitle: stringAllAlbumsByBand){
			for(Item thisItem: items){
				if(thisItem instanceof MusicAlbum){
					if(((MusicAlbum) thisItem).getTitle().equals(thisTitle)){
						itemAllAlbumsByBand.add(thisItem);
					}
				}
			}
		}
		return itemAllAlbumsByBand;
	}
	
	// returns all of the music albums by the specified musician
	public ArrayList<Item> musicByMusician(String musician)
	{
		ArrayList<String> stringAllAlbumsByArtist = new ArrayList<>();
		ArrayList<Item> itemAllAlbumsByArtist = new ArrayList<>();
		
		for(Item item: items){
			if(item instanceof MusicAlbum){
				ArrayList<String> tmpMembers = ((MusicAlbum) item).getMembers();
				if(tmpMembers.contains(musician)){
					stringAllAlbumsByArtist.add(item.getTitle());
				}			
			}
		}
		Collections.sort(stringAllAlbumsByArtist);

		for(String thisTitle: stringAllAlbumsByArtist){
			for(Item thisItem: items){
				if(thisItem instanceof MusicAlbum){
					if(((MusicAlbum) thisItem).getTitle().equals(thisTitle)){
						itemAllAlbumsByArtist.add(thisItem);
					}
				}
			}
		}
		return itemAllAlbumsByArtist;
	}
	
	// returns all of the music albums in the library
	public ArrayList<Item> musicAlbums()
	{
		ArrayList<String> stringAlbums = new ArrayList<>();
		ArrayList<Item> itemAlbums = new ArrayList<>();
		
		for(Item item: items){
			if(item instanceof MusicAlbum){
				stringAlbums.add(item.getTitle());
			}
		}
		
		Collections.sort(stringAlbums);
		
		for(String thisTitle: stringAlbums){
			for(Item thisItem: items){
				if(thisItem instanceof MusicAlbum){
					if(((MusicAlbum) thisItem).getTitle().equals(thisTitle)){
						itemAlbums.add(thisItem);
					}
				}
			}
		}
		return itemAlbums;

	}
	
	// movie-related methods
	
	// adds a movie to the library
	public Item addMovie(String title, String director, int nScenes, String... keywords)
	{
		Item newItem = new Movie(title, director, nScenes, keywords);
		items.add(newItem);
		return newItem;
	}

	// adds the specified actors to a movie
	public void addCast(Item movie, String... members)
	{
		((Movie) movie).setMembers(members);
	}

	// removes a movie from the library
	public boolean removeMovie(String title)
	{
		for(Item item: items){
			if(item instanceof Movie){
				if(item.getTitle().equals(title)){
					items.remove(item);
					return true;
				}
			}
		}
		return false;
	}
	
	// returns all of the movies by the specified director
	public ArrayList<Item> moviesByDirector(String director)
	{
		ArrayList<String> stringMoviesByDirector = new ArrayList<>();
		ArrayList<Item> itemMoviesByDirector = new ArrayList<>();
		
		for(Item item: items){
			if(item instanceof Movie){
				if(((Movie) item).getDirector().equals(director)){
					stringMoviesByDirector.add(item.getTitle());
				}
			}
		}
		Collections.sort(stringMoviesByDirector);

		for(String thisTitle: stringMoviesByDirector){
			for(Item thisItem: items){
				if(thisItem instanceof Movie){
					if(((Movie) thisItem).getTitle().equals(thisTitle)){
						itemMoviesByDirector.add(thisItem);
					}
				}
			}
		}
		return itemMoviesByDirector;
	}
	
	// returns all of the movies by the specified actor
	public ArrayList<Item> moviesByActor(String actor)
	{
		ArrayList<String> stringMoviesByActor = new ArrayList<>();
		ArrayList<Item> itemMoviesByActor = new ArrayList<>();
		
		for(Item item: items){
			if(item instanceof Movie){
				ArrayList<String> tmpActors = ((Movie) item).getCast();
				if(tmpActors.contains(actor)){
					stringMoviesByActor.add(item.getTitle());
				}
			}
		}
		Collections.sort(stringMoviesByActor);

		for(String thisTitle: stringMoviesByActor){
			for(Item thisItem: items){
				if(thisItem instanceof Movie){
					if(((Movie) thisItem).getTitle().equals(thisTitle)){
						itemMoviesByActor.add(thisItem);
					}
				}
			}
		}
		return itemMoviesByActor;
	}
	
	// returns all of the movies in the library
	public ArrayList<Item> movies()
	{
		ArrayList<String> stringAllMovies = new ArrayList<>();
		ArrayList<Item> itemAllMovies = new ArrayList<>();
		for(Item item: items){
			if(item instanceof Movie){
				stringAllMovies.add(item.getTitle());
			}
		}
		Collections.sort(stringAllMovies);
		
		for(String thisTitle: stringAllMovies){
			for(Item thisItem: items){
				if(thisItem instanceof Movie){
					if(((Movie) thisItem).getTitle().equals(thisTitle)){
						itemAllMovies.add(thisItem);
					}
				}
			}
		}
		return itemAllMovies;
	}	
}
